File: android/src/main/java/com/amplitude/android/internal/locators/ComposeLayoutNodeBoundsHelper.java
package com.amplitude.android.internal.locators;

import androidx.annotation.NonNull;
import androidx.compose.ui.geometry.Rect;
import androidx.compose.ui.layout.LayoutCoordinatesKt;
import androidx.compose.ui.node.LayoutNode;
import androidx.compose.ui.node.LayoutNodeLayoutDelegate;

import com.amplitude.common.Logger;

import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Field;

@ApiStatus.Internal
@SuppressWarnings("KotlinInternalInJava")
public class ComposeLayoutNodeBoundsHelper {
  private Field layoutDelegateField = null;

  private final @NotNull Logger logger;

  public ComposeLayoutNodeBoundsHelper(@NonNull Logger logger) {
    this.logger = logger;
    try {
      final Class<?> clazz = Class.forName("androidx.compose.ui.node.LayoutNode");
      layoutDelegateField = clazz.getDeclaredField("layoutDelegate");
      layoutDelegateField.setAccessible(true);
    } catch (Exception e) {
      logger.info("Could not find LayoutNode.layoutDelegate field");
    }
  }

  public @Nullable Rect getLayoutNodeWindowBounds(@NotNull final LayoutNode node) {
    if (layoutDelegateField == null) {
      return null;
    }
    try {
      final LayoutNodeLayoutDelegate delegate = (LayoutNodeLayoutDelegate) layoutDelegateField.get(node);
        if (delegate == null) {
          return null;
        }
        return LayoutCoordinatesKt.boundsInWindow(delegate.getOuterCoordinator().getCoordinates());
    } catch (Exception e) {
      logger.info("Could not fetch position for LayoutNode");
    }
    return null;
  }
}

File: android/src/main/java/com/amplitude/android/internal/locators/ComposeLayoutNodeBoundsHelper.kt
@file:Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE") // to access internal LayoutNode class

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.layout.boundsInWindow
import androidx.compose.ui.node.LayoutNode
import com.amplitude.common.Logger

internal class ComposeLayoutNodeBoundsHelper(private val logger: Logger) {

    private fun getLayoutNodeWindowBounds(node: LayoutNode): Rect? {
        return try {
            // Prefer modifier coordinates if available, otherwise fallback to node coordinates
            val modifierInfo = node.getModifierInfo().firstOrNull()
            val coordinates = modifierInfo?.coordinates ?: node.coordinates
            coordinates.boundsInWindow()
        } catch (e: Exception) {
            logger.warn("Could not fetch position for LayoutNode")
            null
        }
    }

    internal fun layoutNodeBoundsContain(node: LayoutNode, position: Pair<Float, Float>): Boolean {
        val bounds = getLayoutNodeWindowBounds(node) ?: return false
        val (x, y) = position
        return bounds.contains(Offset(x, y))
    }
}
